File: examples/coordinator.ts
// The coordinator seeds work and reads outcomes. It knows nothing about which agent will
// do what — it just posts a `job` and a standalone `task`, then waits for the results to
// appear. Run it against a space that has a planner, workers, and an aggregator running.
//
//   deno run --allow-net --allow-env examples/coordinator.ts

import { RadiaClient, type RadiaRecord } from "../sdk/ts/client.ts";
import { registerDemoKinds } from "./kinds.ts";

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

async function pollFor(fn: () => Promise<RadiaRecord | null>, timeoutMs = 10000): Promise<RadiaRecord | null> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const r = await fn();
    if (r) return r;
    await sleep(150);
  }
  return null;
}

export async function seedAndAwait(client: RadiaClient, text: string): Promise<void> {
  await registerDemoKinds(client);

  // A job to be planned + fanned out, and a standalone task that only worker:reverse matches.
  const job = await client.put({ kind: "job", body: { text } });
  await client.put({ kind: "task", body: { op: "reverse", input: "radia" } });
  console.log(`[coordinator] posted job ${job.id.slice(-6)} ("${text}") + a standalone reverse task`);

  const summary = await pollFor(() => client.readOne({ kind: "summary", match: { jobId: job.id } }));
  if (summary) console.log(`[coordinator] job summary: "${(summary.body as { text: string }).text}"`);
  else console.log(`[coordinator] no summary yet (are the planner/workers/aggregator running?)`);

  const reversed = await pollFor(() => client.readOne({ kind: "result", match: { op: "reverse" } }));
  if (reversed) console.log(`[coordinator] standalone reverse -> "${(reversed.body as { output: string }).output}"`);
}

if (import.meta.main) {
  const client = new RadiaClient();
  await seedAndAwait(client, Deno.args.join(" ") || "the quick brown fox");
}

File: examples/pipeline-py/coordinator.py
"""The coordinator seeds work and reads outcomes, via the extension bindings where they say more
than a raw put: `tools` answers "who is ALIVE to do this" (capabilities policed by presence), and
`seed` writes the record and long-polls for a descendant of the answering kind, so there is no
hand-rolled poll loop here at all.

  python3 examples/pipeline-py/coordinator.py "hello there world"
"""
import sys
import time

from common import PRESENCE_KIND, PRESENCE_TTL_MS, RadiaExtError, connect, register_kinds, require_ext


def live_ops(ext):
    try:
        view = ext.tools(presence_kind=PRESENCE_KIND, ttl_ms=PRESENCE_TTL_MS)
    except RadiaExtError:
        return []  # a fresh space: no worker has declared the capability kind yet
    return sorted(row["tool"] for row in view.get("tools", []))


def wait_for_ops(ext, wanted, timeout_s=10.0):
    """A worker's first beat lands moments after it starts; give the fleet that moment."""
    deadline = time.time() + timeout_s
    while time.time() < deadline:
        ops = live_ops(ext)
        if set(wanted) <= set(ops):
            return ops
        time.sleep(0.3)
    return live_ops(ext)


def seed_and_await(client, ext, text):
    register_kinds(client)

    ops = live_ops(ext)
    print(f"[coordinator] live tools: {', '.join(ops) or '(none: is a worker running?)'}")

    # A job to be planned + fanned out. Its summary lands three links downstream
    # (job -> tasks -> results -> summary), and seed-and-wait answers only a DIRECT child, so the
    # coordinator reads it by CONTENT instead: the summary carries the jobId it answers, which is
    # the same field the aggregator grouped the results by.
    seed_id = ext.seed("pipeline_job", {"text": text})["seedId"]
    summary = None
    deadline = time.time() + 20
    while time.time() < deadline and not summary:
        summary = client.read_one({"kind": "pipeline_summary", "match": {"jobId": seed_id}})
        if not summary:
            time.sleep(0.2)
    if summary:
        print(f'[coordinator] job summary: "{summary["body"]["text"]}"')
    else:
        print("[coordinator] no summary yet (are the planner/workers/aggregator running?)")

    # A standalone task that only worker:reverse matches. Its result IS a direct child (the
    # worker's ack names the task as parent), which is the shape seed-and-wait is for: one call
    # writes the record and long-polls the answer.
    r = ext.seed("pipeline_task", {"op": "reverse", "input": "radia"}, result_kind="pipeline_result", timeout_ms=10_000)
    reversed_ = r.get("result")
    if reversed_:
        print(f'[coordinator] standalone reverse -> "{reversed_["body"]["output"]}"')
    return summary


if __name__ == "__main__":
    client, ext = connect()
    require_ext(ext)
    seed_and_await(client, ext, " ".join(sys.argv[1:]) or "the quick brown fox")
